File: Solution Union Find.java

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.stream.IntStream;

public class Solution {

    public int longestConsecutive(int[] nums) {
        if (nums.length < 2) {
            return nums.length;
        }

        Map<Integer, Integer> valueToIndex = new HashMap<>();
        for (int i = 0; i < nums.length; ++i) {
            valueToIndex.put(nums[i], i);
        }

        UnionFind unionFind = new UnionFind(nums.length);

        for (int n : valueToIndex.keySet()) {
            if (valueToIndex.containsKey(n - 1)) {
                unionFind.joinByRank(unionFind.findParent(valueToIndex.get(n)), unionFind.findParent(valueToIndex.get(n - 1)));
            }
            if (valueToIndex.containsKey(n + 1)) {
                unionFind.joinByRank(unionFind.findParent(valueToIndex.get(n)), unionFind.findParent(valueToIndex.get(n + 1)));
            }
        }

        int[] lengthConsecutiveSequence = new int[nums.length];
        for (int n : valueToIndex.keySet()) {
            ++lengthConsecutiveSequence[unionFind.findParent(valueToIndex.get(n))];
        }

        return Arrays.stream(lengthConsecutiveSequence).max().getAsInt();
    }
}

class UnionFind {

    int[] parent;
    int[] rank;

    UnionFind(int numberOfUniqueElements) {
        parent = IntStream.range(0, numberOfUniqueElements).toArray();
        rank = new int[numberOfUniqueElements];
    }

    int findParent(int index) {
        if (parent[index] != index) {
            parent[index] = findParent(parent[index]);
        }
        return parent[index];
    }

    void joinByRank(int first, int second) {
        if (rank[first] > rank[second]) {
            parent[second] = first;
        } else if (rank[first] < rank[second]) {
            parent[first] = second;
        } else {
            parent[second] = first;
            ++rank[first];
        }
    }
}

File: Solution Union Find.js

/**
 * @param {number[]} nums
 * @return {number}
 */
var longestConsecutive = function (nums) {
    if (nums.length < 2) {
        return nums.length;
    }

    const valueToIndex = new Map();
    for (let i = 0; i < nums.length; ++i) {
        valueToIndex.set(nums[i], i);
    }

    const unionFind = new UnionFind(nums.length);
    for (let [value, index] of valueToIndex) {
        if (valueToIndex.has(value - 1)) {
            unionFind.joinByRank(unionFind.findParent(index), unionFind.findParent(valueToIndex.get(value - 1)));
        }
        if (valueToIndex.has(value + 1)) {
            unionFind.joinByRank(unionFind.findParent(index), unionFind.findParent(valueToIndex.get(value + 1)));
        }
    }

    const lengthConsecutiveSequence = new Array(nums.length).fill(0);
    for (let index of valueToIndex.values()) {
        ++lengthConsecutiveSequence[unionFind.findParent(index)];
    }

    return Math.max(...lengthConsecutiveSequence);
};

class UnionFind {

    constructor(numberOfUniqueElements) {
        this.parent = Array.from(new Array(numberOfUniqueElements).keys());
        this.rank = new Array(numberOfUniqueElements).fill(0);
    }

    findParent(index) {
        if (this.parent[index] !== index) {
            this.parent[index] = this.findParent(this.parent[index]);
        }
        return this.parent[index];
    }

    joinByRank(first, second) {
        if (this.rank[first] > this.rank[second]) {
            this.parent[second] = first;
        } else if (this.rank[first] < this.rank[second]) {
            this.parent[first] = second;
        } else {
            this.parent[second] = first;
            ++this.rank[first];
        }
    }
}
